### PR TITLE
feat(Home): Add impression capping to 'For You'

### PR DESCRIPTION
**Pull request by mmiermans**
# Goal
Show more content that users haven't seen before, by ranking impression capped items lower in the 'For You' slate on Home.

Additional changes:
- Refactor 'For You' rankers to allow impression capping to be slotted in the middle.
- PyCharm changed a .run file after upgrading.

## Reference

Tickets:
* https://getpocket.atlassian.net/browse/HS-206

## Implementation Decisions
- Refactor 'For You' rankers into composable functions.
- Preserve ranking logic, including randomization, to keep this change focused on impression capping.

## QA
**Setup in Pocket-Dev:**
1. Update the 'For You' Feature Group candidate set
2. Set preferred topics for your user to 'Business' and 'Self Improvement'
3. Add impression caps to your user for the top 2 items returned.

**Request using the 'main Pocket-Dev' run configuration:**
```shell
curl --location --request POST 'http://localhost:8000' \
--header 'userId: <user id goes here>' \
--header 'encodedId: <hashed user id goes here>' \
--header 'Content-Type: application/json' \
--data-raw '{
    "query": "query PostmanHomeSlateLineup { homeSlateLineup { id slates { headline moreLink { url text } recommendationReasonType recommendations(count: 20) { corpusItem { id } reason { name type } } } } }",
    "variables": null,
    "operationName": "PostmanHomeSlateLineup"
}'
```

**For You slate from JSON response:**
- [x] The two impression capped items appear after all other preferred topic recs (`1697212b-1065-4e60-839d-c399882a0977` and `357e1f96-617e-4ece-aa21-a9cc74a457a2`)
- [x] Topics spreading is applied
- [x] Preferred topics appear before non-preferred topics
- [x] The order is somewhat random on each request
```json
{
    "headline": "For You",
    "moreLink": null,
    "recommendationReasonType": "PREFERRED_TOPICS",
    "recommendations": [
        {"corpusItem": {"id": "4557a39d-e6f5-46c1-9fcc-09190a3a978b"}, "reason": {"name": "Business", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "f3b97ce4-f614-43a1-8730-e9e0a506cd06"}, "reason": {"name": "Self Improvement", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "105f88de-362b-48c0-bac8-d8d376086bff"}, "reason": {"name": "Business", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "4ac9a337-28a4-499d-a614-a2abc433a0df"}, "reason": {"name": "Self Improvement", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "e429646e-0266-43fb-b3f4-76ddc9a4ccda"}, "reason": {"name": "Business", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "5139c86c-a18f-4cab-9e8a-42a654d7ee4e"}, "reason": {"name": "Self Improvement", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "b39b2f9e-e367-416a-86f8-5bd74ab89e49"}, "reason": {"name": "Business", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "8404d9c6-6f8a-4f61-82fe-ee437209fd74"}, "reason": {"name": "Self Improvement", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "2c8343f2-e708-4ce3-8692-6b0ee2961d44"}, "reason": {"name": "Business", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "6e79df65-763e-4212-8722-765a7a77aeb1"}, "reason": {"name": "Self Improvement", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "1697212b-1065-4e60-839d-c399882a0977"}, "reason": {"name": "Business", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "357e1f96-617e-4ece-aa21-a9cc74a457a2"}, "reason": {"name": "Self Improvement", "type": "PREFERRED_TOPICS"}},
        {"corpusItem": {"id": "3ccfd2a2-ef63-4f60-9070-5a7ec6524dbd"}, "reason": null},
        {"corpusItem": {"id": "d6a02756-49c0-4b6c-99c1-1c526cb6fd83"}, "reason": null},
        {"corpusItem": {"id": "dadd9bba-a41f-4a55-bb9e-3a4883c07b44"}, "reason": null},
        {"corpusItem": {"id": "a485cb61-bbec-4bbc-8d92-900733745280"}, "reason": null},
        {"corpusItem": {"id": "53ca1578-64c7-45de-a7a7-44c64a18739a"}, "reason": null},
        {"corpusItem": {"id": "54157eb4-284d-42b4-8ce6-e0cc412f7f28"}, "reason": null},
        {"corpusItem": {"id": "6d33f864-4588-47a0-ade3-2fe7f6690232"}, "reason": null},
        {"corpusItem": {"id": "9a0930bf-2e3d-4b32-8084-643221360ac9"}, "reason": null}
    ]
},
```

### CODE REVIEW

File: app/data_providers/slate_providers/for_you_slate_provider.py [hunk 1]
```diff
@@ -34,17 +34,24 @@ async def rank_corpus_items(
...
         :return: Ranks items based on topic preferences, with topic spreading.
         """
-        return rank_by_preferred_topics(items, preferred_topics, recommendation_count)
+        # These parameters have default value 'None' to match the signature of the base class, but they are required.
+        assert preferred_topics is not None
+        assert user_impression_capped_list is not None
+
+        items = rank_by_impression_caps(items, user_impression_capped_list)
+        items = spread_topics(items)
+        items = rank_by_preferred_topics(items, preferred_topics)
```
Comment (mmiermans) [inline app/data_providers/slate_providers/for_you_slate_provider.py:53]: @lfishhead This order is different than we had during code pairing. Does this look right to you?

I'll dump a spreadsheet to let Carolyn QA the content.

File: app/data_providers/dispatch.py [hunk 2]
```diff
@@ -59,7 +59,9 @@ async def get_ranked_corpus_slate(self, user: UserIds, recommendation_count: int
             logging.info(f'SetupMoment is unpersonalized for user {user.user_id} because no preferences were found.')
             topics = await self.topic_provider.get_topics(self.DEFAULT_TOPICS)
 
-        items = rank_by_preferred_topics(items, topics, recommendation_count)
+        random.shuffle(items)
+        items = spread_topics(items)
+        items = rank_by_preferred_topics(items, topics)
```
Comment (mmiermans) [inline app/data_providers/dispatch.py:64]: SetupMoment is deprecated and not used, so it doesn't matter what we do here, really. I'll check with Clients and remove this endpoint in a future PR.

**Review by lfishhead — APPROVED** [review]
this was very readable, so much so that I thought i understood it.  great job.

File: app/json/slate_config.schema.json [hunk 3]
```diff
@@ -149,8 +149,7 @@
                   "thompson-sampling-7day",
                   "thompson-sampling-14day",
                   "thompson-sampling-28day",
-                  "firefox-thompson-sampling-1day",
-                  "user-impression-filter"
+                  "firefox-thompson-sampling-1day"
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/json/slate_config.schema.json:152]: That was added as part of the Fx New Tab work we did. It's not used.

**Pull request merged**